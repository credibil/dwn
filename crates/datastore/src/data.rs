//! Data record handling.

use std::collections::BTreeMap;
use std::io::{Cursor, Read, Write};

use anyhow::{Result, anyhow};
use ipld_core::ipld::Ipld;

use crate::BlockStore;
use crate::ipfs::{self, Block};

/// The maximum size of a message.
pub const MAX_ENCODED_SIZE: usize = 30000;

/// The maximum size of a block.
const PARTITION: &str = "DATA";

/// Put a data record into the block store.
pub async fn put(
    owner: &str, record_id: &str, data_cid: &str, reader: impl Read, store: &impl BlockStore,
) -> Result<(String, usize)> {
    ipfs::import(owner, record_id, data_cid, reader, store).await
}

/// Get a data record from the block store.
pub async fn get(
    owner: &str, record_id: &str, data_cid: &str, store: &impl BlockStore,
) -> Result<Option<impl Read>> {
    // get the root block using the partition CID
    let root_cid = root_cid(record_id, data_cid)?;
    let Some(bytes) = store.get(owner, PARTITION, &root_cid).await? else {
        return Ok(None);
    };

    // the root blook contains a list of links to data blocks
    let Ipld::List(links) = ipfs::decode_block(&bytes)? else {
        return Ok(None);
    };

    // TODO: optimize by streaming the data blocks as fetched
    // fetch each data block
    let mut buf = Cursor::new(vec![]);

    for link in links {
        // get data block
        let Ipld::Link(link_cid) = link else {
            return Err(anyhow!("invalid link"));
        };
        let Some(bytes) = store.get(owner, PARTITION, &link_cid.to_string()).await? else {
            return Ok(None);
        };

        // get data block's payload
        let ipld_bytes = ipfs::decode_block(&bytes)?;
        let Ipld::Bytes(bytes) = ipld_bytes else {
            return Ok(None);
        };

        buf.write_all(&bytes)?;
    }

    buf.set_position(0);
    Ok(Some(buf))
}

pub async fn delete(
    owner: &str, record_id: &str, data_cid: &str, store: &impl BlockStore,
) -> Result<()> {
    let root_cid = root_cid(record_id, data_cid)?;
    store.delete(owner, PARTITION, &root_cid).await
}

fn root_cid(record_id: &str, data_cid: &str) -> Result<String> {
    let root = Block::encode(&Ipld::Map(BTreeMap::from([
        (String::from("record_id"), Ipld::String(record_id.to_string())),
        (String::from("data_cid"), Ipld::String(data_cid.to_string())),
    ])))?;
    Ok(root.cid().to_string())
}
