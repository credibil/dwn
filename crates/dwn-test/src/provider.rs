#![allow(missing_docs)]
#![allow(unused_variables)]

//! # Provider
//!
//! Implementation of the `Provider` trait for testing and examples.

pub mod block_store;
pub mod event_log;
mod event_stream;
pub mod key_store;
pub mod message_store;
pub mod task_store;

use anyhow::{Result, anyhow};
use blockstore::InMemoryBlockstore;
use surrealdb::Surreal;
use surrealdb::engine::local::{Db, Mem};
use vercre_dwn_server::provider::{DidResolver, Document, Provider};

use self::key_store::{ALICE_DID, KeyStoreImpl};

const NAMESPACE: &str = "integration-test";

#[derive(Clone)]
pub struct ProviderImpl {
    db: Surreal<Db>,
    blockstore: InMemoryBlockstore<64>,
    pub nats_client: async_nats::Client,
    pub keystore: KeyStoreImpl,
}

impl Provider for ProviderImpl {}

impl ProviderImpl {
    pub async fn new() -> Result<Self> {
        let db = Surreal::new::<Mem>(()).await?;
        db.use_ns(NAMESPACE).use_db(ALICE_DID).await?;

        Ok(Self {
            db,
            blockstore: InMemoryBlockstore::<64>::new(),
            nats_client: async_nats::connect("demo.nats.io").await?,
            keystore: KeyStoreImpl::new(),
        })
    }
}

impl DidResolver for ProviderImpl {
    async fn resolve(&self, url: &str) -> Result<Document> {
        if url == ALICE_DID {
            return serde_json::from_slice(include_bytes!("./provider/data/alice_did.json"))
                .map_err(|e| anyhow!(format!("issue deserializing document: {e}")));
        } else {
            return serde_json::from_slice(include_bytes!("./provider/data/bob_did.json"))
                .map_err(|e| anyhow!(format!("issue deserializing document: {e}")));
        }
    }
}
