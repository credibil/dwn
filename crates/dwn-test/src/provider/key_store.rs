//! # Keystore

use std::collections::HashMap;

use anyhow::{Result, anyhow};
use base64ct::{Base64UrlUnpadded, Encoding};
use ed25519_dalek::{Signer as _, SigningKey};
use vercre_dwn::provider::{KeyStore, Keyring};
use vercre_infosec::{Algorithm, PublicKey, Receiver, SecretKey, SharedSecret, Signer};

// use x25519_dalek::{PublicKey, StaticSecret};
use crate::provider::ProviderImpl;

pub const ALICE_DID: &str = "did:key:z6Mkj8Jr1rg3YjVWWhg7ahEYJibqhjBgZt1pDCbT4Lv7D4HX";
const ALICE_SECRET_KEY: &str = "8rmFFiUcTjjrL5mgBzWykaH39D64VD0mbDHwILvsu30";

pub const BOB_DID: &str = "did:key:z6MkqWGVUwMwt4ahxESTVg1gjvxZ4w4KkXomksSMdCB3eHeD";
const BOB_SECRET_KEY: &str = "n8Rcm64tLob0nveDUuXzP-CnLmn3V11vRqk6E3FuKCo";

pub const CAROL_DID: &str = "did:key:z6MkuY2MjELw3xQExptJtVkuW5YSjfeQYrVZv41RMYrwZmYd";
const CAROL_SECRET_KEY: &str = "V0YsmES1Tc8-sozoyYeBKemcrUaOLq_IceWbjWwmbMo";

pub const APP_DID: &str = "did:key:z6Mkj85hWKz3rvxVt6gL54rCsEMia8ZRXMTmxaUv4yLDSnTA";
const APP_SECRET_KEY: &str = "fAe8yt4xBaDpyuPKY9_1NBxmiFMCfVnnryMXD-oLyVk";

pub const INVALID_DID: &str = "did:key:z6Mkj85hWKz3rvxVt6gL54rCsEMia8ZRXMTmxaUv4yLDSnTA";
const INVALID_SECRET_KEY: &str = "n8Rcm64tLob0nveDUuXzP-CnLmn3V11vRqk6E3FuKCo";

#[derive(Default, Clone, Debug)]
pub struct KeyStoreImpl {
    keyrings: HashMap<String, KeyringImpl>,
}

#[derive(Default, Clone, Debug)]
pub struct KeyringImpl {
    pub did: String,
    pub secret_key: String,
}

impl KeyStoreImpl {
    // Populate the keystore with test keyrings
    pub fn new() -> Self {
        let mut keyrings = HashMap::new();

        let alice_keyring = KeyringImpl {
            did: ALICE_DID.to_string(),
            secret_key: ALICE_SECRET_KEY.to_string(),
        };
        let bob_keyring = KeyringImpl {
            did: BOB_DID.to_string(),
            secret_key: BOB_SECRET_KEY.to_string(),
        };
        let carol_keyring = KeyringImpl {
            did: CAROL_DID.to_string(),
            secret_key: CAROL_SECRET_KEY.to_string(),
        };
        let app_keyring = KeyringImpl {
            did: APP_DID.to_string(),
            secret_key: APP_SECRET_KEY.to_string(),
        };
        let invalid_keyring = KeyringImpl {
            did: INVALID_DID.to_string(),
            secret_key: INVALID_SECRET_KEY.to_string(),
        };

        keyrings.insert(ALICE_DID.to_string(), alice_keyring);
        keyrings.insert(BOB_DID.to_string(), bob_keyring);
        keyrings.insert(CAROL_DID.to_string(), carol_keyring);
        keyrings.insert(APP_DID.to_string(), app_keyring);
        keyrings.insert(INVALID_DID.to_string(), invalid_keyring);

        Self { keyrings }
    }
}

impl KeyStore for ProviderImpl {
    fn keyring(&self, identifier: &str) -> Result<impl Keyring> {
        self.keystore.keyrings.get(identifier).cloned().ok_or_else(|| anyhow!("keyring not found"))
    }
}

impl Keyring for KeyringImpl {}

impl Signer for KeyringImpl {
    async fn try_sign(&self, msg: &[u8]) -> Result<Vec<u8>> {
        let decoded = Base64UrlUnpadded::decode_vec(&self.secret_key)?;
        let secret_key: ed25519_dalek::SecretKey =
            decoded.try_into().map_err(|_| anyhow!("invalid secret key"))?;
        let signing_key: SigningKey = SigningKey::from_bytes(&secret_key);

        Ok(signing_key.sign(msg).to_bytes().to_vec())
    }

    async fn public_key(&self) -> Result<Vec<u8>> {
        let decoded = Base64UrlUnpadded::decode_vec(&self.secret_key)?;
        let secret_key: ed25519_dalek::SecretKey =
            decoded.try_into().map_err(|_| anyhow!("invalid secret key"))?;
        let signing_key: SigningKey = SigningKey::from_bytes(&secret_key);

        Ok(signing_key.verifying_key().as_bytes().to_vec())
    }

    fn algorithm(&self) -> Algorithm {
        Algorithm::EdDSA
    }

    async fn verification_method(&self) -> Result<String> {
        let verify_key = self.did.strip_prefix("did:key:").unwrap_or_default();
        Ok(format!("{}#{}", self.did, verify_key))
    }
}

impl Receiver for KeyringImpl {
    // // FIXME: wrap return in Result<Vec<u8>>
    // fn public_key(&self) -> Vec<u8> {
    //     let decoded = Base64UrlUnpadded::decode_vec(&self.secret_key).unwrap();
    //     let bytes: [u8; 32] =
    //         decoded.try_into().map_err(|_| anyhow!("Invalid secret key")).unwrap();

    //     // derive x25519 key agreement public key
    //     PublicKey::from(bytes).as_bytes().to_vec()
    // }

    fn key_id(&self) -> String {
        self.did.clone()
    }

    async fn shared_secret(&self, sender_public: PublicKey) -> Result<SharedSecret> {
        let decoded = Base64UrlUnpadded::decode_vec(&self.secret_key)?;
        let bytes: [u8; 32] = decoded.try_into().map_err(|_| anyhow!("invalid secret key"))?;

        let secret_key = SecretKey::from(bytes);
        secret_key.shared_secret(sender_public)
    }
}
