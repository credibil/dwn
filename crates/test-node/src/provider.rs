#![allow(missing_docs)]
#![allow(unused_variables)]

//! # Provider
//!
//! Implementation of the `Provider` trait for testing and examples.

pub mod block_store;
mod event_stream;
pub mod key_store;

use anyhow::Result;
use blockstore::InMemoryBlockstore;
use dwn_node::provider::{
    DataStore, DidResolver, Document, EventLog, MessageStore, Provider, TaskStore,
};

#[derive(Clone)]
pub struct ProviderImpl {
    blockstore: InMemoryBlockstore<64>,
    pub nats_client: async_nats::Client,
}

impl ProviderImpl {
    pub async fn new() -> Result<Self> {
        Ok(Self {
            blockstore: InMemoryBlockstore::<64>::new(),
            nats_client: async_nats::connect("demo.nats.io").await?,
        })
    }
}

impl Provider for ProviderImpl {}
impl MessageStore for ProviderImpl {}
impl DataStore for ProviderImpl {}
impl EventLog for ProviderImpl {}
impl TaskStore for ProviderImpl {}

impl DidResolver for ProviderImpl {
    async fn resolve(&self, url: &str) -> Result<Document> {
        // let resolved =
        //     vercre_did::resolve(url, None, self.clone()).await.map_err(|e| anyhow!(e))?;
        // Ok(resolved.document.unwrap())

        // if url == &alice.did {
        //     serde_json::from_slice(include_bytes!("./provider/data/alice_did.json"))
        //         .map_err(|e| anyhow!(format!("issue deserializing document: {e}")))
        // } else {
        //     serde_json::from_slice(include_bytes!("./provider/data/bob_did.json"))
        //         .map_err(|e| anyhow!(format!("issue deserializing document: {e}")))
        // }

        println!("resolve: {}", url);

        Ok(Document::default())
    }
}
