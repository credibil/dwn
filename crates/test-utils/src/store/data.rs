use std::future::Future;
use std::io::Read;

use vercre_dwn::provider::DataStore;

use super::ProviderImpl;

impl DataStore for ProviderImpl {
    fn put(
        &self, tenant: &str, record_id: &str, data_cid: &str, data: impl Read,
    ) -> impl Future<Output = anyhow::Result<()>> + Send {
        async { Ok(()) }
    }

    fn get(
        &self, tenant: &str, record_id: &str, data_cid: &str,
    ) -> impl Future<Output = anyhow::Result<Option<impl Read>>> + Send {
        let buf = vec![];
        let reader = std::io::Cursor::new(buf);
        async { Ok(Some(reader)) }
    }

    async fn delete(&self, tenant: &str, record_id: &str, data_cid: &str) -> anyhow::Result<()> {
        todo!()
    }

    async fn purge(&self) -> anyhow::Result<()> {
        todo!()
    }
}
