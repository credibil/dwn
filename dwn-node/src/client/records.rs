//! # Records Query

use anyhow::{Result, anyhow};
use chrono::{DateTime, Utc};

use crate::authorization::AuthorizationBuilder;
use crate::data::cid;
use crate::provider::Signer;
use crate::records::{Delete, DeleteDescriptor};
use crate::{Descriptor, Interface, Method};

/// Options to use when creating a permission grant.
#[derive(Clone, Debug, Default)]
pub struct DeleteBuilder {
    message_timestamp: DateTime<Utc>,
    record_id: Option<String>,
    prune: Option<bool>,
    permission_grant_id: Option<String>,
    protocol_role: Option<String>,
}

impl DeleteBuilder {
    /// Returns a new [`DeleteBuilder`]
    #[must_use]
    pub fn new() -> Self {
        Self {
            message_timestamp: Utc::now(),
            ..Self::default()
        }
    }

    /// Specifies the permission grant ID.
    #[must_use]
    pub fn record_id(mut self, record_id: impl Into<String>) -> Self {
        self.record_id = Some(record_id.into());
        self
    }

    /// Specifies the permission grant ID.
    #[must_use]
    pub const fn prune(mut self, prune: bool) -> Self {
        self.prune = Some(prune);
        self
    }

    /// Specifies the permission grant ID.
    #[must_use]
    pub fn permission_grant_id(mut self, permission_grant_id: impl Into<String>) -> Self {
        self.permission_grant_id = Some(permission_grant_id.into());
        self
    }

    /// Specifies the permission grant ID.
    #[must_use]
    pub fn protocol_role(mut self, protocol_role: impl Into<String>) -> Self {
        self.protocol_role = Some(protocol_role.into());
        self
    }

    /// Build the write message.
    ///
    /// # Errors
    /// LATER: Add errors
    pub async fn build(self, signer: &impl Signer) -> Result<Delete> {
        let Some(record_id) = self.record_id else {
            return Err(anyhow!("`record_id` is not set"));
        };

        let descriptor = DeleteDescriptor {
            base: Descriptor {
                interface: Interface::Records,
                method: Method::Delete,
                message_timestamp: self.message_timestamp,
            },
            record_id,
            prune: self.prune.unwrap_or(false),
        };

        let mut auth_builder =
            AuthorizationBuilder::new().descriptor_cid(cid::from_value(&descriptor)?);
        if let Some(id) = self.permission_grant_id {
            auth_builder = auth_builder.permission_grant_id(id);
        }
        if let Some(role) = self.protocol_role {
            auth_builder = auth_builder.protocol_role(role);
        }
        let authorization = auth_builder.build(signer).await?;

        Ok(Delete {
            descriptor,
            authorization,
        })
    }
}
