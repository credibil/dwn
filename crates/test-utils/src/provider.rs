//! # Mock Provider

use std::sync::Arc;

use anyhow::Result;
use blockstore::{Blockstore as _, InMemoryBlockstore};
use cid::Cid;
use credibil_core::api::Client;
use credibil_dwn::event::{Event, Subscriber};
use credibil_dwn::provider::{BlockStore, EventStream, Resolver};
use credibil_proof::DocumentRequest;
use futures::stream::StreamExt;
use multihash_codetable::MultihashDigest;
use serde::{Deserialize, Serialize};

use crate::store::Store;

const RAW: u64 = 0x55;
const SUBJECT: &str = "events";
#[derive(Clone)]
pub struct Provider {
    blockstore: Arc<InMemoryBlockstore<64>>,
    pub nats_client: Arc<async_nats::Client>,
}

impl Provider {
    pub async fn new() -> Self {
        Self {
            blockstore: Arc::new(InMemoryBlockstore::<64>::new()),
            nats_client: Arc::new(
                async_nats::connect("demo.nats.io").await.expect("should connect"),
            ),
        }
    }
}

impl BlockStore for Provider {
    async fn put(&self, owner: &str, partition: &str, key: &str, block: &[u8]) -> Result<()> {
        let cid = Identitifier::new(owner, partition, key).to_cid()?;
        self.blockstore.put_keyed(&cid, block).await.map_err(Into::into)
    }

    async fn get(&self, owner: &str, partition: &str, key: &str) -> Result<Option<Vec<u8>>> {
        let cid = Identitifier::new(owner, partition, key).to_cid()?;
        let Some(bytes) = self.blockstore.get(&cid).await? else {
            return Ok(None);
        };
        Ok(Some(bytes))
    }

    async fn delete(&self, owner: &str, partition: &str, key: &str) -> Result<()> {
        let cid = Identitifier::new(owner, partition, key).to_cid()?;
        self.blockstore.remove(&cid).await?;
        Ok(())
    }

    async fn purge(&self, _owner: &str, _partition: &str) -> Result<()> {
        unimplemented!()
    }
}

#[derive(Serialize, Deserialize)]
struct Identitifier<'a> {
    owner: &'a str,
    partition: &'a str,
    key: &'a str,
}

impl<'a> Identitifier<'a> {
    fn new(owner: &'a str, partition: &'a str, key: &'a str) -> Self {
        Self {
            owner,
            partition,
            key,
        }
    }

    fn to_cid(&self) -> anyhow::Result<Cid> {
        let mut buf = Vec::new();
        ciborium::into_writer(self, &mut buf)?;
        let hash = multihash_codetable::Code::Sha2_256.digest(&buf);
        Ok(Cid::new_v1(RAW, hash))
    }
}

impl Resolver for Provider {
    async fn resolve(&self, url: &str) -> Result<Vec<u8>> {
        let request = DocumentRequest { url: url.to_string() };
        let document = Client::new(Store)
            .request(request)
            .owner("owner")
            .execute()
            .await
            .map(|r| r.0.clone())?;
        serde_json::to_vec(&document).map_err(|e| e.into())
    }
}

impl EventStream for Provider {
    /// Subscribe to a owner's event stream.
    async fn subscribe(&self, owner: &str) -> Result<Subscriber> {
        let subscriber = self.nats_client.subscribe(format!("{SUBJECT}.{owner}")).await?;
        let mapped = subscriber.map(|m| serde_json::from_slice::<Event>(&m.payload).unwrap());
        Ok(Subscriber::new(mapped))
    }

    /// Emits an event to a owner's event stream.
    async fn emit(&self, owner: &str, event: &Event) -> Result<()> {
        let bytes = serde_json::to_vec(event)?;
        self.nats_client.publish(format!("{SUBJECT}.{owner}"), bytes.into()).await?;
        Ok(())
    }
}
