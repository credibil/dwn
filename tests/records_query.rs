//! Messages Subscribe

use dwn_test::key_store::{ALICE_DID, BOB_DID};
use dwn_test::provider::ProviderImpl;
use http::StatusCode;
use insta::assert_yaml_snapshot as assert_snapshot;
use rand::RngCore;
use vercre_dwn::data::{DataStream, MAX_ENCODED_SIZE};
use vercre_dwn::provider::KeyStore;
use vercre_dwn::records::{QueryBuilder, RecordsFilter, Sort, Write, WriteBuilder, WriteData};
use vercre_dwn::{Error, endpoint};

// Should return a status of BadRequest (400) when querying for unpublished records
// with sort date set to `Sort::Publishedxxx`.
#[tokio::test]
async fn invalid_sort() {
    let provider = ProviderImpl::new().await.expect("should create provider");
    let alice_keyring = provider.keyring(ALICE_DID).expect("should get Alice's keyring");

    let mut query = QueryBuilder::new()
        .filter(RecordsFilter::new().published(false))
        .sign(&alice_keyring)
        .build()
        .await
        .expect("should create query");

    query.descriptor.date_sort = Some(Sort::PublishedAscending);
    let Err(Error::BadRequest(e)) = endpoint::handle(ALICE_DID, query.clone(), &provider).await
    else {
        panic!("should be BadRequest");
    };
    assert_eq!(e, "cannot sort by `date_published` when querying for unpublished records");

    query.descriptor.date_sort = Some(Sort::PublishedDescending);
    let Err(Error::BadRequest(e)) = endpoint::handle(ALICE_DID, query, &provider).await else {
        panic!("should be BadRequest");
    };
    assert_eq!(e, "cannot sort by `date_published` when querying for unpublished records");
}

// Should return `record_id`, `descriptor`, `authorization` and `attestation` fields.
#[tokio::test]
async fn return_values() {
    let provider = ProviderImpl::new().await.expect("should create provider");
    let alice_keyring = provider.keyring(ALICE_DID).expect("should get Alice's keyring");
    let bob_keyring = provider.keyring(BOB_DID).expect("should get Bob's keyring");

    // --------------------------------------------------
    // Alice writes a record.
    // --------------------------------------------------
    let stream = DataStream::from(br#"{"message": "test record write"}"#.to_vec());

    let write = Write::build()
        .data(WriteData::Reader(stream.clone()))
        .data_format("awesome_data_format")
        .attest(&[&bob_keyring])
        .sign(&alice_keyring)
        .build()
        .await
        .expect("should create write");
    let reply = endpoint::handle(ALICE_DID, write.clone(), &provider).await.expect("should write");
    assert_eq!(reply.status.code, StatusCode::ACCEPTED);

    // --------------------------------------------------
    // Alice queries for records with matching format.
    // --------------------------------------------------
    let filter = RecordsFilter::new().add_author(ALICE_DID).data_format("awesome_data_format");
    let query = QueryBuilder::new()
        .filter(filter)
        .sign(&alice_keyring)
        .build()
        .await
        .expect("should create query");
    let reply = endpoint::handle(ALICE_DID, query, &provider).await.expect("should query");
    assert_eq!(reply.status.code, StatusCode::OK);

    let query_reply = reply.body.expect("should have reply");
    let entries = query_reply.entries.expect("should have entries");
    assert_eq!(entries.len(), 1);

    assert_snapshot!("return_values", entries[0].write, {
        ".recordId" => "[recordId]",
        ".descriptor.messageTimestamp" => "[messageTimestamp]",
        ".descriptor.dateCreated" => "[dateCreated]",
        ".authorization.signature.payload" => "[payload]",
        ".authorization.signature.signatures[0].signature" => "[signature]",
        ".attestation.payload" => "[payload]",
        ".attestation.signatures[0].signature" => "[signature]",
    });
}

// Should return matching records.
#[tokio::test]
async fn find_matches() {
    let provider = ProviderImpl::new().await.expect("should create provider");
    let alice_keyring = provider.keyring(ALICE_DID).expect("should get Alice's keyring");

    // --------------------------------------------------
    // Alice writes 3 records.
    // --------------------------------------------------
    let stream = DataStream::from(br#"{"message": "test record write"}"#.to_vec());

    for i in 1..=3 {
        let mut builder = WriteBuilder::new().data(WriteData::Reader(stream.clone()));

        if i > 1 {
            builder = builder.data_format("awesome_data_format").schema(format!("schema_{i}"));
        }

        let write = builder.sign(&alice_keyring).build().await.expect("should create write");
        let reply = endpoint::handle(ALICE_DID, write, &provider).await.expect("should write");
        assert_eq!(reply.status.code, StatusCode::ACCEPTED);
    }

    // --------------------------------------------------
    // Alice queries for records with matching format.
    // --------------------------------------------------
    let query = QueryBuilder::new()
        .filter(RecordsFilter::new().data_format("awesome_data_format"))
        .sign(&alice_keyring)
        .build()
        .await
        .expect("should create query");
    let reply = endpoint::handle(ALICE_DID, query, &provider).await.expect("should query");
    assert_eq!(reply.status.code, StatusCode::OK);

    let query_reply = reply.body.expect("should have reply");
    let entries = query_reply.entries.expect("should have entries");
    assert_eq!(entries.len(), 2);

    // --------------------------------------------------
    // Alice queries for records with matching schema.
    // --------------------------------------------------
    let query = QueryBuilder::new()
        .filter(RecordsFilter::new().data_format("awesome_data_format").schema("schema_2"))
        .sign(&alice_keyring)
        .build()
        .await
        .expect("should create query");
    let reply = endpoint::handle(ALICE_DID, query, &provider).await.expect("should query");
    assert_eq!(reply.status.code, StatusCode::OK);

    let query_reply = reply.body.expect("should have reply");
    let entries = query_reply.entries.expect("should have entries");
    assert_eq!(entries.len(), 1);
}

// Should return `encoded_data` if data size is within the spec threshold.
#[tokio::test]
async fn encoded_data() {
    let provider = ProviderImpl::new().await.expect("should create provider");
    let alice_keyring = provider.keyring(ALICE_DID).expect("should get Alice's keyring");

    // --------------------------------------------------
    // Alice writes a record.
    // --------------------------------------------------
    let stream = DataStream::from(br#"{"message": "test record write"}"#.to_vec());

    let write = Write::build()
        .data(WriteData::Reader(stream.clone()))
        .sign(&alice_keyring)
        .build()
        .await
        .expect("should create write");
    let reply = endpoint::handle(ALICE_DID, write.clone(), &provider).await.expect("should write");
    assert_eq!(reply.status.code, StatusCode::ACCEPTED);

    // --------------------------------------------------
    // Alice queries for record, expecting to get `encoded_data`.
    // --------------------------------------------------
    let query = QueryBuilder::new()
        .filter(RecordsFilter::new().record_id(write.record_id))
        .sign(&alice_keyring)
        .build()
        .await
        .expect("should create query");
    let reply = endpoint::handle(ALICE_DID, query, &provider).await.expect("should query");
    assert_eq!(reply.status.code, StatusCode::OK);

    let query_reply = reply.body.expect("should have reply");
    let entries = query_reply.entries.expect("should have entries");
    let entry = &entries[0];
    assert!(entry.write.encoded_data.is_some());
}

// Should return `encoded_data` if data size is within the spec threshold.
#[tokio::test]
async fn no_encoded_data() {
    let provider = ProviderImpl::new().await.expect("should create provider");
    let alice_keyring = provider.keyring(ALICE_DID).expect("should get Alice's keyring");

    // --------------------------------------------------
    // Alice writes a record.
    // --------------------------------------------------
    let mut data = [0u8; MAX_ENCODED_SIZE + 10];
    rand::thread_rng().fill_bytes(&mut data);
    let stream = DataStream::from(data.to_vec());

    let write = Write::build()
        .data(WriteData::Reader(stream.clone()))
        .sign(&alice_keyring)
        .build()
        .await
        .expect("should create write");
    let reply = endpoint::handle(ALICE_DID, write.clone(), &provider).await.expect("should write");
    assert_eq!(reply.status.code, StatusCode::ACCEPTED);

    // --------------------------------------------------
    // Alice queries for record, expecting to get `encoded_data`.
    // --------------------------------------------------
    let query = QueryBuilder::new()
        .filter(RecordsFilter::new().record_id(write.record_id))
        .sign(&alice_keyring)
        .build()
        .await
        .expect("should create query");
    let reply = endpoint::handle(ALICE_DID, query, &provider).await.expect("should query");
    assert_eq!(reply.status.code, StatusCode::OK);

    let query_reply = reply.body.expect("should have reply");
    let entries = query_reply.entries.expect("should have entries");
    let entry = &entries[0];
    assert!(entry.write.encoded_data.is_none());
}
