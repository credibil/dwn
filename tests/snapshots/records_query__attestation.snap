---
source: tests/records_query.rs
assertion_line: 1965
expression: "entries[0].write.attestation"
snapshot_kind: text
---
payload: "[payload]"
signatures:
  - protected: eyJhbGciOiJFZERTQSIsInR5cCI6Imp3dCIsImtpZCI6ImRpZDprZXk6ejZNa2o4SnIxcmczWWpWV1doZzdhaEVZSmlicWhqQmdadDFwRENiVDRMdjdENEhYI3o2TWtqOEpyMXJnM1lqVldXaGc3YWhFWUppYnFoakJnWnQxcERDYlQ0THY3RDRIWCJ9
    signature: "[signature]"
